### PR TITLE
StdOut publisher

### PR DESCRIPTION
**Pull request by hoegaarden**
This is a spike exploring the proposal layed out in https://github.com/kbialek/deye-inverter-mqtt/issues/89

### CODE REVIEW

File: src/deye_stdout_publisher.py
Comment (hoegaarden) [reply src/deye_stdout_publisher.py:39]: Don't think so.
I'd leave it to the user and the tool they use to figure out what they want / need on stderr / stdout.
While I can't think of a good use-case myself, users might want to pipe both logs and data into the same thing and split / process them there.

**Comment by kbialek** [comment]
Included in beta image `ghcr.io/kbialek/deye-inverter-mqtt:2023.06.1-beta-2`

**Comment by hoegaarden** [comment]
Let me know if you are happy with that PR, especially looking for feedback on:
- naming of the thing and its config knob: `DeyeStdoutPublisher`, `DEYE_FEATURE_STDOUT_PUBLISHER`, ...
- general implementation, integration & testing
- docs

File: README.md
Comment (kbialek) [inline README.md:161]: Please add this env var in `config.env.example` too

File: README.md
Comment (kbialek) [inline README.md:195]: This file is already too large. Would you mind to shorten this section and move the details into a new file under `docs` folder?

**Review by kbialek — CHANGES_REQUESTED** [review]

**Comment by kbialek** [comment]
> Let me know if you are happy with that PR, especially looking for feedback on:
> 
>     * naming of the thing and its config knob: `DeyeStdoutPublisher`, `DEYE_FEATURE_STDOUT_PUBLISHER`, ...

okay

> 
>     * general implementation, integration & testing

okay

> 
>     * docs

comments left

File: README.md
Comment (hoegaarden) [reply README.md]: Done.
Took the liberty to also include the other feature gates.

File: README.md
Comment (hoegaarden) [reply README.md]: Spit the stdout publisher docs into its own file, doc'ing the telegraf example there too.

File: docs/stdout_publisher.md
Comment (hoegaarden) [inline docs/stdout_publisher.md:68]: @kbialek I switched the telegraf example to run `deye-inverter-mqtt` as a container. This would only work if you plan to publish the new version of the container image with the `latest` tag, which I think you do, right?

Comment (kbialek) [reply docs/stdout_publisher.md:68]: Yes, the most recent release is tagged as `latest`

File: docs/stdout_publisher.md
Comment (kbialek) [inline docs/stdout_publisher.md:23]: What is the benefit of having these properties in the output?
If it really has to stay, I'd rename `sensor` to `sensor_type`

File: docs/stdout_publisher.md
Comment (kbialek) [inline docs/stdout_publisher.md:24]: Imho it would make more sense to rename this property to `sensor`. WDYT?

File: docs/stdout_publisher.md
Comment (kbialek) [inline docs/stdout_publisher.md:19]: The value property name is not consistent across metrics. Here it's `temp`, next it's `energy`. Wouldn't it be simpler to use constant property name, e.g. `value`?

File: docs/stdout_publisher.md
Comment (hoegaarden) [reply docs/stdout_publisher.md:23]: I want to expose "everything" I can reasonably get and leave it to the user/consumer to ignore or collect/process it.

Renaming to `sensor_type`, tho, probably makes sense.

File: docs/stdout_publisher.md
Comment (hoegaarden) [reply docs/stdout_publisher.md:19]: This plays very nice with using telegraf to collect and process those metrics.

telegraf, with the [json_v2 parser](https://github.com/influxdata/telegraf/tree/master/plugins/parsers/json_v2), will consider (pretty much) everything that is not used as [a `timestamp`](https://github.com/kbialek/deye-inverter-mqtt/pull/90/files#diff-008cb0b3f78bb87f24168135b73bc2aba1a25d00eafb89467386ec1d2d942755R85) or as [a `tag`](https://github.com/kbialek/deye-inverter-mqtt/pull/90/files#diff-008cb0b3f78bb87f24168135b73bc2aba1a25d00eafb89467386ec1d2d942755R87) as a `field` (= "metric value").

Essentially the `field` in combination with [the `measurement_name`](https://github.com/kbialek/deye-inverter-mqtt/pull/90/files#diff-008cb0b3f78bb87f24168135b73bc2aba1a25d00eafb89467386ec1d2d942755R76) builds up the metric name. So that leaves me with the following metrics / time-series when pushed to prometheus:
```
deye_up{serial="123123"} 1
deye_energy{serial="123123", source="dc/pv1", unit="W", ...} 123
deye_energy{serial="123123", source="dc/pv2", unit="W", ...} 99
deye_current{serial="123123", source="dc/pv1", unit="A", ...} 2.1
deye_current{serial="123123", source="dc/pv2", unit="A", ...} 1.2
...
```
So one time series per "dimension" (energy, up, current, ...), with different tags (telegraf speak) = labels (prometheus speak).

If we'd use the a generic `value` key in the JSON object, we'd probably also need a `name`/`metric_name`/... key. Then, for my use-case, I'd probably need a processor step, which transforms from e.g. `deye_value{serial="123123", metric_name="energy", ...} 0.3` to `deye_energy{serial="123123", ...} 0.3`.

Now, I am not generally against using a generic `value` field.

I wouldn't do it, however, because
- this works nice with the use-case at hand
- I don't know of other use-cases
- I would rather "optimize" for the use-case I know of
    - rather than for a hypothetical, yet unknown, use-case
    - having said that, of course this should not block hypothetical future use-cases either
        - which it doesn't ; other consumers, if that data-model does not fit them, can still transform the data as needed


Again, not against it, just laying out my thoughts and reasoning.

Comment (kbialek) [reply docs/stdout_publisher.md:19]: Okay, then perhaps this publisher should be renamed to `TelegrafJsonV2Publisher`. WDYT?

Comment (hoegaarden) [reply docs/stdout_publisher.md:19]: I am not sure, that's a mouth full, TBH, and too specific to telegraf. The output can still be consumed by anything that can parse json, not just telegraf. And that's definitely one of my goals. It "just happens" to work really nice with telegraf/json_v2 out of the box.

What do you think about
- keeping the name as is
- explicitly call out in the docs that
    - this can be used with anything that can parse json
    - however, it uses a format that works really well with telegraf/json_v2
    - if that format does not work for specific use-cases, ask users to open issues

If we find real good evidence that another format is needed, we can implement another config knob (e.g. `STDOUT_FORMAT`) in a backwards compatible way. I am not saying we should, just that we could, if the need should arise.

File: docs/stdout_publisher.md
Comment (hoegaarden) [reply docs/stdout_publisher.md]: Renamed `sensor` -> `sensor_type`

File: docs/stdout_publisher.md
Comment (hoegaarden) [reply docs/stdout_publisher.md]: Renamed `source` -> `sensor`

File: docs/stdout_publisher.md
Comment (kbialek) [reply docs/stdout_publisher.md:19]: Let me ask differently. Who is backing this format? Is it Influx? https://docs.influxdata.com/telegraf/v1.21/data_formats/input/json_v2/

Chosen name should be specific and self explanatory.

> If we find real good evidence that another format is needed, we can implement another config knob (e.g. STDOUT_FORMAT) in a backwards compatible way. I am not saying we should, just that we could, if the need should arise.

This is an illusive future requirement. Don't take it personally. Imho, devs have a tendency to anticipate such imagined features and as a result drag unnecessary complexity into the projects. Therefore, if such a requirement materialized, my approach would be to implement a new publisher, and provide a migration path. And as for now, I strongly prefer to keep it possibly simple.

Comment (hoegaarden) [reply docs/stdout_publisher.md:19]: > Let me ask differently. Who is backing this format? Is it Influx?

I guess a combination of the implementation of json_v2 in telegraf and [gjson](https://github.com/tidwall/gjson/). However, it's able to consume various different layouts of json documents. Further down you can see an example with a different input format and a changed telegraf config which would still yield the same data.

The point I am trying to make is: It's just json. Yes, it plays nice with telegraf/json_v2, but it is still telegraf would be capable of consuming different layouts of the json document on the one hand, and any other tool, say `jq` for example, would be able to consume and make sense of the currently proposed layout too. Thus I am not convinced to call it `TelegrafJsonV2Publisher` as neither telegraf nor json_v2 really define any specifics, but can be configured to work with various different layouts. Therefore I argue this is a 'generic layout' that any tool that is capable of parsing json should be able to consume and make sense of.

Having said that, in the end I don't care how it's called. You make the call.

But coming back to the "main point":
Would you rather prefer objects like:
```json
{
  "value": 23,   # <- the key is always the generic "value"
  "metric": "temp",  # <- the actual "metric name" is in another field
  "unit": "something",
    ...
}
```
or
```json
{
  "temp": 23,   # <-- the key is the "metric name"
  "unit": "something",
    ...
}
```
I'd prefer the later, which is what the current implementation implements. But I think you lean towards the former, is that right?

> This is an illusive future requirement. Don't take it personally. Imho, devs have a tendency to anticipate such imagined features and as a result drag unnecessary complexity into the projects. Therefore, if such a requirement materialized, my approach would be to implement a new publisher, and provide a migration path. And as for now, I strongly prefer to keep it possibly simple.

I think we are on the same page here, generally.
- I don't see a need to implement any other output format, the current one should be able to handle anything if the user's tool is capable of parsing json. Any transformation into a destination system's specifics needs to be handled by the consumer anyway.
- If I am wrong, and there is a need for a different layout / format this can still be implemented. Nothing that this PR introduces would block that; that's pretty much the only point I tried to make
- If a hypothetical future output format would be implemented in `DeyeStdoutPublisher` or it's own thing, how it would be configured, ... that's all implementation details IMHO.

<details>
<summary>Example with different json object layout</summary>

If `DeyeStdoutPublisher` would produce slightly different output like
```json
{
  "serial": "1234567890",
  "address": "192.168.0.1",
  "port": 8899,
  "data": [
    {
      "value": 1,
      "metric": "up"
    }, {
      "value": 31,
      "metric": "temp",
      "name": "Radiator temperature",
      "unit": "°C",
      "groups": "micro",
      "sensor_type": "SingleRegisterSensor",
      "sensor": "radiator",
      "timestamp": 1686219130
    }, {
      "value": 0.2,
      "metric": "energy",
      "name": "PV1 Total",
      "unit": "kWh",
      "groups": "micro",
      "sensor_type": "DoubleRegisterSensor",
      "sensor": "dc/pv1/total",
      "timestamp": 1686219130
    }, {
      ...
    }
  ]
}
```
telegraf with json_v2 would still be able to consume this with e.g. a config something like that:
```toml
[[inputs.file]]
  files = ["new.json"]
  data_format = "json_v2"

  [[inputs.file.json_v2]]
    measurement_name = "deye"

    [[inputs.file.json_v2.object]]
      path = "data"

      timestamp_key = "timestamp"
      timestamp_format = "unix"

      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="power")#.value'
        rename = "power"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="energy")#.value'
        rename = "energy"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="up")#.value'
        rename = "up"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="temp")#.value'
        rename = "temp"

      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor'
      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor_type'
      [[inputs.file.json_v2.object.tag]]
        path = '#.unit'
      [[inputs.file.json_v2.object.tag]]
        path = '#.groups'
      [[inputs.file.json_v2.object.tag]]
        path = '#.name'

    [[inputs.file.json_v2.tag]]
      path = 'address'
    [[inputs.file.json_v2.tag]]
      path = 'port'
    [[inputs.file.json_v2.tag]]
      path = 'serial'
```

Note: I assume, this can  be made way nicer, if one understands how gjson really works, which I don't.

</details>

Comment (hoegaarden) [reply docs/stdout_publisher.md:19]: > Let me ask differently. Who is backing this format? Is it Influx?

I guess a combination of the implementation of json_v2 in telegraf and [gjson](https://github.com/tidwall/gjson/). However, it's able to consume various different layouts of json documents. Further down you can see an example with a different input format and a changed telegraf config which would still yield the same data.

The point I am trying to make is: It's just json. Yes, it plays nice with telegraf/json_v2, but it is still telegraf would be capable of consuming different layouts of the json document on the one hand, and any other tool, say `jq` for example, would be able to consume and make sense of the currently proposed layout too. Thus I am not convinced to call it `TelegrafJsonV2Publisher` as neither telegraf nor json_v2 really define any specifics, but can be configured to work with various different layouts. Therefore I argue this is a 'generic layout' that any tool that is capable of parsing json should be able to consume and make sense of.

Having said that, in the end I don't care how it's called. You make the call.

But coming back to the "main point":
Would you rather prefer objects like:
```json
{
  "value": 23,   # <- the key is always the generic "value"
  "metric": "temp",  # <- the actual "metric name" is in another field
  "unit": "something",
    ...
}
```
or
```json
{
  "temp": 23,   # <-- the key is the "metric name"
  "unit": "something",
    ...
}
```
I'd prefer the later, which is what the current implementation implements. But I think you lean towards the former, is that right?

> This is an illusive future requirement. Don't take it personally. Imho, devs have a tendency to anticipate such imagined features and as a result drag unnecessary complexity into the projects. Therefore, if such a requirement materialized, my approach would be to implement a new publisher, and provide a migration path. And as for now, I strongly prefer to keep it possibly simple.

I think we are on the same page here, generally.
- I don't see a need to implement any other output format, the current one should be able to handle anything if the user's tool is capable of parsing json. Any transformation into a destination system's specifics needs to be handled by the consumer anyway.
- If I am wrong, and there is a need for a different layout / format this can still be implemented. Nothing that this PR introduces would block that; that's pretty much the only point I tried to make
- If a hypothetical future output format would be implemented in `DeyeStdoutPublisher` or it's own thing, how it would be configured, ... that's all implementation details IMHO.

<details>
<summary>Example with different json object layout</summary>

If `DeyeStdoutPublisher` would produce slightly different output like
```json
{
  "serial": "1234567890",
  "address": "192.168.0.1",
  "port": 8899,
  "data": [
    {
      "value": 1,
      "metric": "up"
    }, {
      "value": 31,
      "metric": "temp",
      "name": "Radiator temperature",
      "unit": "°C",
      "groups": "micro",
      "sensor_type": "SingleRegisterSensor",
      "sensor": "radiator",
      "timestamp": 1686219130
    }, {
      "value": 0.2,
      "metric": "energy",
      "name": "PV1 Total",
      "unit": "kWh",
      "groups": "micro",
      "sensor_type": "DoubleRegisterSensor",
      "sensor": "dc/pv1/total",
      "timestamp": 1686219130
    }, {
      ...
    }
  ]
}
```
telegraf with json_v2 would still be able to consume this with e.g. a config something like that:
```toml
[[inputs.file]]
  files = ["new.json"]
  data_format = "json_v2"

  [[inputs.file.json_v2]]
    measurement_name = "deye"

    [[inputs.file.json_v2.object]]
      path = "data"

      timestamp_key = "timestamp"
      timestamp_format = "unix"

      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="power")#.value'
        rename = "power"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="energy")#.value'
        rename = "energy"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="up")#.value'
        rename = "up"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="temp")#.value'
        rename = "temp"

      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor'
      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor_type'
      [[inputs.file.json_v2.object.tag]]
        path = '#.unit'
      [[inputs.file.json_v2.object.tag]]
        path = '#.groups'
      [[inputs.file.json_v2.object.tag]]
        path = '#.name'

    [[inputs.file.json_v2.tag]]
      path = 'address'
    [[inputs.file.json_v2.tag]]
      path = 'port'
    [[inputs.file.json_v2.tag]]
      path = 'serial'
```

Note: I assume, this can  be made way nicer, if one understands how gjson really works, which I don't.

</details>

Comment (hoegaarden) [reply docs/stdout_publisher.md:19]: > Let me ask differently. Who is backing this format? Is it Influx?

I guess a combination of the implementation of json_v2 in telegraf and [gjson](https://github.com/tidwall/gjson/). However, it's able to consume various different layouts of json documents. Further down you can see an example with a different input format and a changed telegraf config which would still yield the same data.

The point I am trying to make is: It's just json. Yes, it plays nice with telegraf/json_v2, but it is still telegraf would be capable of consuming different layouts of the json document on the one hand, and any other tool, say `jq` for example, would be able to consume and make sense of the currently proposed layout too. Thus I am not convinced to call it `TelegrafJsonV2Publisher` as neither telegraf nor json_v2 really define any specifics, but can be configured to work with various different layouts. Therefore I argue this is a 'generic layout' that any tool that is capable of parsing json should be able to consume and make sense of.

Having said that, in the end I don't care how it's called. You make the call.

But coming back to the "main point":
Would you rather prefer objects like:
```json
{
  "value": 23,   # <- the key is always the generic "value"
  "metric": "temp",  # <- the actual "metric name" is in another field
  "unit": "something",
    ...
}
```
or
```json
{
  "temp": 23,   # <-- the key is the "metric name"
  "unit": "something",
    ...
}
```
I'd prefer the later, which is what the current implementation implements. But I think you lean towards the former, is that right?

> This is an illusive future requirement. Don't take it personally. Imho, devs have a tendency to anticipate such imagined features and as a result drag unnecessary complexity into the projects. Therefore, if such a requirement materialized, my approach would be to implement a new publisher, and provide a migration path. And as for now, I strongly prefer to keep it possibly simple.

I think we are on the same page here, generally.
- I don't see a need to implement any other output format, the current one should be able to handle anything if the user's tool is capable of parsing json. Any transformation into a destination system's specifics needs to be handled by the consumer anyway.
- If I am wrong, and there is a need for a different layout / format this can still be implemented. Nothing that this PR introduces would block that; that's pretty much the only point I tried to make
- If a hypothetical future output format would be implemented in `DeyeStdoutPublisher` or it's own thing, how it would be configured, ... that's all implementation details IMHO.

<details>
<summary>Example with different json object layout</summary>

If `DeyeStdoutPublisher` would produce slightly different output like
```json
{
  "serial": "1234567890",
  "address": "192.168.0.1",
  "port": 8899,
  "data": [
    {
      "value": 1,
      "metric": "up"
    }, {
      "value": 31,
      "metric": "temp",
      "name": "Radiator temperature",
      "unit": "°C",
      "groups": "micro",
      "sensor_type": "SingleRegisterSensor",
      "sensor": "radiator",
      "timestamp": 1686219130
    }, {
      "value": 0.2,
      "metric": "energy",
      "name": "PV1 Total",
      "unit": "kWh",
      "groups": "micro",
      "sensor_type": "DoubleRegisterSensor",
      "sensor": "dc/pv1/total",
      "timestamp": 1686219130
    }, {
      ...
    }
  ]
}
```
telegraf with json_v2 would still be able to consume this with e.g. a config something like that:
```toml
[[inputs.file]]
  files = ["new.json"]
  data_format = "json_v2"

  [[inputs.file.json_v2]]
    measurement_name = "deye"

    [[inputs.file.json_v2.object]]
      path = "data"

      timestamp_key = "timestamp"
      timestamp_format = "unix"

      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="power")#.value'
        rename = "power"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="energy")#.value'
        rename = "energy"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="up")#.value'
        rename = "up"
      [[inputs.file.json_v2.object.field]]
        path = '#(metric=="temp")#.value'
        rename = "temp"

      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor'
      [[inputs.file.json_v2.object.tag]]
        path = '#.sensor_type'
      [[inputs.file.json_v2.object.tag]]
        path = '#.unit'
      [[inputs.file.json_v2.object.tag]]
        path = '#.groups'
      [[inputs.file.json_v2.object.tag]]
        path = '#.name'

    [[inputs.file.json_v2.tag]]
      path = 'address'
    [[inputs.file.json_v2.tag]]
      path = 'port'
    [[inputs.file.json_v2.tag]]
      path = 'serial'
```

Note: I assume, this can  be made way nicer, if one understands how gjson really works, which I don't.

</details>

Comment (kbialek) [reply docs/stdout_publisher.md:19]: Thank you for your elaborated answers. After reading that I came to the following conclusion:
1. `json_v2` is not a data format, but a parser spec used by the data consumer. So it does not matter much how the output JSON will look like, as the client can still be configured to digest it. Hence, further discussions about the JSON format are pointless.
2. The functionality that you've implemented does not fit into the scope of this project. IIRC, I've already told that before when you came with the idea, and now I'm very confident about that. I prefer to follow the Unix philosophy of doing one thing and doing it well. On the other hand this same philosophy also promotes extensibility.
3. So, I'd like to help you to achieve your goal. I've decided to add a simple plugin system that will allow you to do what you want, as you like, without asking anyone for approval. I encourage you to give it a try and transform your code into the plugin. See https://github.com/kbialek/deye-inverter-mqtt/tree/feature/plugin-loader IMO, this change will let us keep the things maintainable and properly decoupled. 
4. In case you want to make your work public, then start a github project. I will be happy to link it in the readme.

Comment (hoegaarden) [reply docs/stdout_publisher.md:19]: Ok then.
Thanks for the plugin mechanism!

Comment (kbialek) [reply docs/stdout_publisher.md:19]: > Thanks for the plugin mechanism!

No problem. Please let me know if it works for you as expected.

Comment (hoegaarden) [reply docs/stdout_publisher.md:19]: Yeah, I am running with `ghcr.io/hoegaarden/deye-inverter-mqtt-plugins:55db5c1` from [here](https://github.com/hoegaarden/deye-inverter-mqtt-plugins/) -- works for me.

**Pull request closed without merging**
